File: cpp/v2.0.2022/prime_logging.cpp
#include <iostream>
#include <cmath>
#include <chrono>
#include <string>
#include <fstream>

using namespace std;

int main()
{
    const int scope[] = {100, 1000, 10000, 100000, 1000000, 10000000, 25000000, 100000000, 1000000000};
    const int reference[] = {25, 168, 1229, 9592, 78498, 664579, 1565927, 5761455, 50847534};
    for (int i = 0; i < 9; i++)
    {
        cout << scope[i] << " ";
        int last = scope[i];
        int found = 4;   // we already know 2, 3, 5, 7
        const int arraylength = (int)(last / log(last));
        int primes[10000] = {3, 5, 7};
        cout << "Calculating prime numbers until " << last << " in C++ v2.0.2022" << endl;
        auto start = chrono::high_resolution_clock::now();
        for (int number = 11; number < last; number += 2)
        {
            int prime = 1;
            for (int divider = 3; divider < (int)(sqrt(number)) + 1; divider += 2)
            {
                if (number % divider == 0)
                {
                    prime = 0;
                    break;
                }
            }
            if (prime ==  1)
            {
                found += 1;
            }
        }
        auto end = chrono::high_resolution_clock::now();
        double time_taken = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
        time_taken *= 1e-9;
        cout << "Found " << found << " prime numbers. It should be ";
        cout << reference[i] << " primes." << endl;
        cout << "This took " << fixed << time_taken << " seconds." << endl << endl;
        string filename = std::to_string(scope[i]) + ".txt";
        ofstream logfile(filename);
        logfile << "Calculating prime numbers until " << last  << " in C++ v2.0.2022" << endl;
        logfile << "Found " << found << " prime numbers. It should be ";
        logfile << reference[i] << " primes." << endl;
        logfile << "This took " << fixed << time_taken << " seconds." << endl << endl;
        logfile.close();
    }
}

File: python/v2.0.2022/prime_logging.py
# prime v2.0.2022 with modulo operator 
# cycles through limits and writes to the filesystem

import math, time, os, cpuinfo

scope = [100, 1000, 10000, 100000, 1000000, 10000000, 25000000, 100000000, 1000000000]
reference = [25, 168, 1229, 9592, 78498, 664579, 1565927, 5761455, 50847534]

if __name__ == "__main__":
    for i in range(len(scope)):
        last = scope[i]
        found = 4              # we start from 11, know 2, 3, 5, 7
        primes = [3, 5, 7]     # exclude 2 since we only test odd numbers    
        print(f"Prime numbers to {last} in Python with algorithm v2.0.2022")
        start = time.perf_counter_ns()
        for number in range(11, last, 2):
            prime = True
            for divider in range(3, int(math.sqrt(number))+1, 2):
                if number % divider == 0:
                    prime = False
                    break
            if prime:
                found += 1
        end = time.perf_counter_ns()
        print(f"This took: {(end - start)/1000} microseconds.")
        print(f"I found {found} prime numbers. Should be {reference[i]}.\n")
        filename = str(last) + ".txt"
        with open(filename, "w") as fp:
            fp.write('prime v2.0.2022 in python\n')            
            fp.write(cpuinfo.get_cpu_info()['brand_raw'])
            fp.write(f'\nPrimes to {last} took {(end - start)/1000} microseconds.')
            fp.write(f'\nFound {found} primes. Should be {reference[i]}.\n')
